### PR TITLE
fix krb5_msg_type: Unable to detect AS-REQ, TGS-REQ and KRB-ERROR

### PR DESCRIPTION
**Pull request by zer1t0**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4529

Related SV test: https://github.com/OISF/suricata-verify/pull/572

The problem is that the krb5_msg_type doesn't trigger properly for:
- AS-REQ (krb5_msg_type: 10): This rule matches the KRB-ERROR message that follows an AS-REQ, which is not the expected behaviour. 
- TGS-REQ (krb5_msg_type: 12): Same behaviour as AS-REQ.
-  KRB-ERROR (krb5_msg_type: 30) It's not triggered at all.

To solve the problem the following actions were taken in `KRB5State.parse`:
- Add transaction when parsing AS-REQ message
- Add transaction when parsing TGS-REQ message
- Change the message type of KRB-ERROR to MessageType::KRB_ERROR

suricata-verify-pr: 572

### CODE REVIEW

File: rust/src/krb/log.rs
Comment (catenacyber) [inline rust/src/krb/log.rs:31]: How can we get here ?

File: rust/src/krb/krb5.rs
Comment (catenacyber) [inline rust/src/krb/krb5.rs:149]: I do not know much about Kerberos.
What does this 12 mean ?
Any link to a RFC ?

File: rust/src/krb/krb5.rs
Comment (catenacyber) [inline rust/src/krb/krb5.rs:135]: Could you please add a comment about this ?
Why using the condition `if self.req_id > 0`
My understanding is that what got parsed as a request id, is not in this case a request id but a request type...
Why did we parse it as a request id in the first place ?

File: rust/src/krb/krb5.rs
Comment (catenacyber) [inline rust/src/krb/krb5.rs:190]: Why only if `self.req_id > 0` ?

**Review by catenacyber — COMMENTED** [review]
Thanks for this work.
I left some questions

File: rust/src/krb/krb5.rs
Comment (zer1t0) [reply rust/src/krb/krb5.rs:135]: Hi,

You are totally correct, the req_id is the request type, however that was the name of the variable in the code. I wanted to do as few changes in the code as I could so I keep this variable name (I also needed to introduce a field to keep track of request_type in responses, so I already use the word `req_type`, more accurate). But yeah, it requires a rename, 
 probably in a improvement PR.

Related to the condition `if self.req_id > 0`: in the code self.req_id is only set different to zero for requests, so you only get the request type when the previous one was a request, in other case, you keep the default None Option.

File: rust/src/krb/krb5.rs
Comment (zer1t0) [reply rust/src/krb/krb5.rs:190]: in the code self.req_id is only set different to zero for requests, so you only get the request type when the previous one was a request, in other case, you keep the default None Option.

File: rust/src/krb/log.rs
Comment (zer1t0) [reply rust/src/krb/log.rs:31]: You arrive to this conditional when a KRB-ERROR message is logged, if the previous request type is set, then you return the request type. However, in case of the request type being None (because start capturing in the middle of an UDP connection, for example) you return the UNKNOWN value.

**Comment by zer1t0** [comment]
> 

Hi, I try to answer your questions. Keep in mind that this is a fix request so I tried to do just fix changes, without varying the style of the code. I think you point some code improvement issues that requires consideration, but those require another kind of PR.

File: rust/src/krb/log.rs
Comment (catenacyber) [reply rust/src/krb/log.rs:31]: Thanks. Maybe this is worth a comment in the code

File: rust/src/krb/krb5.rs
Comment (catenacyber) [reply rust/src/krb/krb5.rs:149]: Could you create a redmine ticket to track this improvement ?

File: rust/src/krb/krb5.rs
Comment (catenacyber) [reply rust/src/krb/krb5.rs:135]: Thanks for the explanation. This naming deserves a comment

File: rust/src/krb/krb5.rs
Comment (catenacyber) [reply rust/src/krb/krb5.rs:190]: ok

File: rust/src/krb/krb5.rs
Comment (zer1t0) [reply rust/src/krb/krb5.rs:151]: Here it is https://redmine.openinfosecfoundation.org/issues/4950

File: rust/src/krb/krb5.rs
Comment (zer1t0) [reply rust/src/krb/krb5.rs:137]: Done in https://github.com/OISF/suricata/pull/6508/commits/abbef13b12b58e8b6c2fb7fff6b589d97cfacde5

File: rust/src/krb/log.rs
Comment (zer1t0) [reply rust/src/krb/log.rs:34]: Done in https://github.com/OISF/suricata/pull/6508/commits/4642ecf72cf88f99df193b1a029f0f2ffb17850a

File: rust/src/krb/krb5.rs
Comment (catenacyber) [inline rust/src/krb/krb5.rs:124]: Is not this redundant ?

Comment (catenacyber) [reply rust/src/krb/krb5.rs:124]: I mean is `tx.etype` not None before that ?

**Comment by catenacyber** [comment]
Changes look good to me...
I guess some commits can be squashed together for the final version
Let's wait for CI

**Comment by catenacyber** [comment]
Do you understand why CI fails with suricate-verify ?

**Comment by zer1t0** [comment]
> Do you understand why CI fails with suricate-verify ?

It is failing in `krb5-request-frag-log` test with error:
```
===> krb5-request-frag-log: Sub test #1: FAIL : expected 2 matches; got 4 for filter {'count': 2, 'match': {'event_type': 'alert', 'alert.signature_id': 1}}
===> krb5-request-frag-log: Sub test #2: FAIL : expected 2 matches; got 4 for filter {'count': 2, 'match': {'event_type': 'alert', 'alert.signature_id': 2}}
```

It seems these rules are matching many more times than expected:
```
alert krb5 any any -> any any (krb5.sname; content:"test"; sid:1;)
alert krb5 any any -> any any (krb5.cname; content:"user01"; sid:2;)
```

That is normal with the fix, since now these rules are also matching with kerberos requests (AS-REQ and TGS-REQ), so the number of matches increases.

File: rust/src/krb/krb5.rs
Comment (zer1t0) [reply rust/src/krb/krb5.rs:124]: Yes, you are right, it is None previously too.

**Comment by catenacyber** [comment]
Ok, so could you change the S-V test so that they the right number of matches ?

**Comment by catenacyber** [comment]
And thanks for the explanation :-)

**Comment by zer1t0** [comment]
> Ok, so could you change the S-V test so that they the right number of matches ?

Done in https://github.com/OISF/suricata-verify/pull/572/commits/f02f87fdd5b6fb811a9de6f437fb2472830510c5

**Comment by catenacyber** [comment]
Rerunning CI...

**Comment by catenacyber** [comment]
@jasonish how do you feel about backports for this ?
I guess this is a feature addition and we do not want to back port it.
But then, we will need the Suricata-verify test to tell 2 alerts up to version 6, and 4 alerts from version 7 onwards...

**Comment by jasonish** [comment]
> @jasonish how do you feel about backports for this ? I guess this is a feature addition and we do not want to back port it. But then, we will need the Suricata-verify test to tell 2 alerts up to version 6, and 4 alerts from version 7 onwards...

This reads more like a bug than a feature to me.

**Comment by catenacyber** [comment]
Said like that, it may sound like a bug, but this is because we were only parsing half the messages... So ?

**Comment by zer1t0** [comment]
It is a bug since it is expected to match the message you are referring to in the `krb5_msg_type`, but this doesn't happen in half of the cases, not even in the krb5_msg_type [documentation](https://suricata.readthedocs.io/en/suricata-6.0.4/rules/kerberos-keywords.html#krb5-msg-type) example.

**Review by catenacyber — APPROVED** [review]
Looks good to me.

@jasonish do you know what is the process to merge this ?
That is this fix needs to be merged in master, master6, master5 and Suricata-verify.
The CI will be red when we land one fix until we merge the other ones

**Review by victorjulien — CHANGES_REQUESTED** [review]
Can you update the git commits to prefix them with `krb5`?

**Comment by victorjulien** [comment]
> Looks good to me.
> 
> @jasonish do you know what is the process to merge this ? That is this fix needs to be merged in master, master6, master5 and Suricata-verify. The CI will be red when we land one fix until we merge the other ones

wrt backports I would like to understand better how much more log volume this will generate for the new transactions that are generated.

**Comment by zer1t0** [comment]
> > Looks good to me.
> > @jasonish do you know what is the process to merge this ? That is this fix needs to be merged in master, master6, master5 and Suricata-verify. The CI will be red when we land one fix until we merge the other ones
> 
> wrt backports I would like to understand better how much more log volume this will generate for the new transactions that are generated.

Right now only the responses generate a transaction. With the fix, also the requests will generate a transaction, so we would have twice as many transactions. Since Kerberos is just a authentication protocol where each connection usually exchanges between 2 and 4 messages (2 requests and 2 responses), it shouldn't affect negatively to the performance.

**Comment by victorjulien** [comment]
> > > Looks good to me.
> > > @jasonish do you know what is the process to merge this ? That is this fix needs to be merged in master, master6, master5 and Suricata-verify. The CI will be red when we land one fix until we merge the other ones
> > 
> > 
> > wrt backports I would like to understand better how much more log volume this will generate for the new transactions that are generated.
> 
> Right now only the responses generate a transaction. With the fix, also the requests will generate a transaction, so we would have twice as many transactions. Since Kerberos is just a authentication protocol where each connection usually exchanges between 2 and 4 messages (2 requests and 2 responses), it shouldn't affect negatively to the performance.

I've found a pcap where I have real krb5 traffic. In master it generates 12k records, in this branch 24k. So this is significant increase in log volume. I don't think we should do that in a backport. For master its fine.

**Comment by victorjulien** [comment]
Btw please submit a new PR against master when implementing changes.

**Comment by zer1t0** [comment]
New PR: https://github.com/OISF/suricata/pull/6926

**Pull request closed without merging**
